### PR TITLE
Implement support for parameterized numpy types - np.ndarray[int] etc

### PR DESCRIPTION
**Pull request by jobh**
A follow-up on https://github.com/HypothesisWorks/hypothesis/pull/3668, addressing https://github.com/HypothesisWorks/hypothesis/issues/3150 (draft, for discussion).

I believe something like this is the easiest way to support arbitrary and parameterized special types such as `np.ndarray[int]`. Allowing the numpy extra to inspect the type before other partial resolutions means we can compare directly with e.g. `np.typing.ArrayLike`, which is (IMO) much easier than digesting the inner types of the union. Mind, if the union is not consumed by the specialized resolver then it will get another crack at the inner types later.

I did try to make it sort-of-generic by adding a lookup table for specialized resolvers. The lookup table only considers installed modules though, not the actual module of the type.

### CODE REVIEW

**Comment by jobh** [comment]
I see now that there are multiple PRs for this already, I didn't read the bug report properly. Sorry if just adding confusion. But hey, at least we're mapping out the solution space!

**Comment by Zac-HD** [comment]
No worries, the previous PRs were closed mostly just because the work stalled out so a new one is entirely welcome!

I think I'd still prefer to avoid handling `ArrayLike` separately from the inner types - it's more robust to people defining their own unions, for example - but the approach otherwise looks good on a quick skim.  Once tests are passing, or skipped on versions of Numpy which predate `np.typing`, I'll do a deeper review; or sooner if you request one 🙂

**Comment by jobh** [comment]
> I think I'd still prefer to avoid handling `ArrayLike` separately from the inner types - it's more robust to people defining their own unions, for example - but the approach otherwise looks good on a quick skim. Once tests are passing, or skipped on versions of Numpy which predate `np.typing`, I'll do a deeper review; or sooner if you request one slightly_smiling_face

No need for a review just yet, feedback on the approach is what I'm after so far (appreciated!).

It would be best to handle ArrayLike normally, yes. But its constituents are private types like `np._typing._ArrayLike._SupportsArray` and `np._typing._nested_sequence._NestedSequence` (both of type `typing._ProtocolMeta`), so my fear is that we either end up relying on brittle implementation details or have to basically implement a complete shadow type system.

My opinion is that handling `ArrayLike` as a composite type is the lesser evil, but yes: it should not be too difficult to handle `_NestedSequence` and `_SupportsArray` in a rudimentary way insted.

**Comment by Zac-HD** [comment]
> Handling `ArrayLike` as a composite type is the lesser evil, but it should not be too difficult to handle `_NestedSequence` and `_SupportsArray` in a rudimentary way instead.  Do you still prefer the decomposed approach, even using private types?

Weak preference for decomposition because I suspect that some Numpy functions will use `_SupportsArray` and it would be nice to have that work well with e.g. the Ghostwriter and introspection generally, but I'll trust your judgement if you think that composed handling is still better all-things-considered.

**Comment by jobh** [comment]
So, I tried to decompose. The main problem is that `ArrayLike` is too loose: it accepts (and hence generates) values that are not accepted by `np.array`, contrary to the contract. This would be good for the numpy project if they used hypothesis, but bad for numpy end users.

`ArrayLike` is (with `base_types := [bool, int, float, complex, str, bytes]`):
```
typing.Union[
  numpy._typing._array_like._SupportsArray[numpy.dtype[typing.Any]],
  numpy._typing._nested_sequence._NestedSequence[
    numpy._typing._array_like._SupportsArray[numpy.dtype[typing.Any]]   (*)
  ],
  *base_types,
  numpy._typing._nested_sequence._NestedSequence[
    typing.Union[
      *base_types                                                       (**)
    ]
  ]
]
```

At `(*)`, equal-shape arrays are not enforced, and at `(**)` some combinations of types fail (minimal example: `['', b'\x80']`)

So: I'll resurrect the composite handling of ArrayLike, but if you want I can keep the decomposed support as well (in case e.g. `_SupportsArray` is used elsewhere)

**Comment by jobh** [comment]
@Zac-HD It is ready for review now, not sure if tests are 100% yet but mostly.

**Comment by jobh** [comment]
> So: I'll resurrect the composite handling of ArrayLike, but if you want I can keep the decomposed support as well (in case e.g. _SupportsArray is used elsewhere)

I added a test for a duplicate of `ArrayLike` as well, which is not handled as composite.

File: hypothesis-python/RELEASE.rst
Comment (Zac-HD) [inline hypothesis-python/RELEASE.rst:5]: Style guide for release notes:

- I like to assume that readers are smart, but haven't heard of <whatever I'm writing about> - so adding cross-references to the docs for `numpy.typing.{ArrayLike,DtypeLike,NDArray}` would be nice.  Thanks to [intersphinx](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html) (😉), this can use standard cross-reference syntax!
- Avoid mentioning implementation details or non-public interfaces - in this case, early resolution and the scaffolding should not be end-user-visible.

(looks like [our guide doc](https://github.com/HypothesisWorks/hypothesis/blob/master/guides/documentation.rst#changelog-entries) could have this added, I'll do that soon)

File: hypothesis-python/src/hypothesis/extra/numpy.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/extra/numpy.py:54]: I think this would be a cleaner interface if it took `(module_name, attr_name)` and did an `assert "." not in attr_name` to clarify that it has to be a top-level name.

File: hypothesis-python/src/hypothesis/extra/numpy.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/extra/numpy.py:1006]: I'm not a fan of interfaces that catch _any_ error and move on; it's too easy to miss unexpected errors.  Can we return `None` for all cases we can't infer?  Or if not, restrict the expected error types?

(the answer may well be "it's not worth the effort in this case", e.g. because it's version-dependent)

File: hypothesis-python/src/hypothesis/extra/numpy.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/extra/numpy.py:1009]: I'd move these out as top-level functions, with an `_`-prefix - it's a slight performance optimization, but useful as it clarifies that we're not accessing nonlocal variables from the enclosing function.

File: hypothesis-python/src/hypothesis/extra/numpy.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/extra/numpy.py:1066]: I'd probably just assign this instead of having the function above, and then write additional `base_strat_except_bytes` _and_ `base_strat_except_str` (though unfortunately covering both will take yet more code)

File: hypothesis-python/src/hypothesis/strategies/_internal/types.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/strategies/_internal/types.py:603]:
```suggestion
_global_extra_lookup: typing.Dict[str, typing.Callable[[type], Optional[st.SearchStrategy]]] = {
```
adding `from typing import ...` at the top would probably be good for readability here.

File: hypothesis-python/src/hypothesis/strategies/_internal/types.py
Comment (Zac-HD) [inline hypothesis-python/src/hypothesis/strategies/_internal/types.py:606]: I think I'd prefer a non-lambda function here - this is getting moderately complicated, and if it fails the traceback wouldn't be amazing.  At that point we can also do
```python
def _from_numpy_type(thing):
    from hypothesis.extra.numpy import _from_type

    _global_extra_lookup["numpy"] = _from_type
    return _from_type(thing)
```
for slightly faster / simpler subsequent calls.

File: hypothesis-python/tests/numpy/test_from_type.py
Comment (Zac-HD) [inline hypothesis-python/tests/numpy/test_from_type.py:70]: Might get more concise if we have a global `needs_np_typing = {"reason": "numpy.typing is not available"}` to `**`unpack in each decorator.

File: hypothesis-python/tests/numpy/test_from_type.py
Comment (Zac-HD) [inline hypothesis-python/tests/numpy/test_from_type.py:166]: Is this specifically that _bytes doesn't work_, or _we can't mix str and bytes_?  Either is fine, I'm just keen to keep an accurate record of the problems you hit because I expect it to come up later.

**Review by Zac-HD — COMMENTED** [review]
Looks great again - thanks @jobh!

I've got a bunch of pretty fiddly feedback below, but the structure and logic of it all seems ready so this is just docs, style, and clarification stuff.

File: hypothesis-python/src/hypothesis/extra/numpy.py
Comment (jobh) [reply hypothesis-python/src/hypothesis/extra/numpy.py:54]: Will do. Should it be moved to some common helpers module, or leave it where it is until there is another caller?

File: hypothesis-python/src/hypothesis/extra/numpy.py
Comment (jobh) [reply hypothesis-python/src/hypothesis/extra/numpy.py:1006]: No problem, it is in fact the docstring that is wrong here (the callsite does not consume exceptions)

File: hypothesis-python/src/hypothesis/extra/numpy.py
Comment (jobh) [reply hypothesis-python/src/hypothesis/extra/numpy.py:1066]: I'll do that. Actually I was a bit uncertain whether it makes a difference to re-use the same strategy object in multiple places (`one_of(integers(), integers())` vs `i=integers; one_of(i, i)` — I'd think "no", and your comment seems to validate this?

File: hypothesis-python/src/hypothesis/strategies/_internal/types.py
Comment (jobh) [reply hypothesis-python/src/hypothesis/strategies/_internal/types.py:603]: I usually leave `typing.` in, but that's mostly habit. But: do you mean changing all the (many) references to `typing.` in that file, or just `Optional/Dict/Callable`? If the latter IMO it's better to be consistent within the file even if it's a bit more verbose.

File: hypothesis-python/src/hypothesis/extra/numpy.py
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/extra/numpy.py:54]: I'd leave it here.

File: hypothesis-python/src/hypothesis/extra/numpy.py
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/extra/numpy.py:1066]: Correct, strategies are all cached so reusing vs "re-creating" is in fact almost the same operation (both just fancy dict lookups)

File: hypothesis-python/src/hypothesis/strategies/_internal/types.py
Comment (Zac-HD) [reply hypothesis-python/src/hypothesis/strategies/_internal/types.py:603]: Ugh, yeah, in this particular file we should probably leave them all as `typing.` - that's distinguishing between `typing`, `typing_extensions`, and possibly `collections.abc` in places 😅

File: hypothesis-python/tests/numpy/test_from_type.py
Comment (jobh) [reply hypothesis-python/tests/numpy/test_from_type.py:166]: The latter. `[b'\x80', '']` fails to convert to `ndarray` with some ascii-decode-error. Probably a numpy bug. I added a comment-link to the PR that shows this example, but I can inline it instead

File: hypothesis-python/src/hypothesis/strategies/_internal/types.py
Comment (jobh) [reply hypothesis-python/src/hypothesis/strategies/_internal/types.py:606]: And in fact it could even remove itself from `_global_extra_lookup` on `ImportError`!

File: hypothesis-python/tests/numpy/test_from_type.py
Comment (Zac-HD) [reply hypothesis-python/tests/numpy/test_from_type.py:166]: Oh dear, this looks pretty nasty: if you pass a mixture of str and bytes, Numpy will try to decode all the bytes as ascii-encoded strings - so that's the minimal non-ascii-decodable bytestring, and the minimal unicode string.

Arguably mixing strings and bytes like this should be an upstream error, but... `np.array(["unicode", 1.337])` also casts everything to unicode strings, so I guess the problem is only the assumption that bytestrings must be ascii-encoded.

How about we just limit our strategy to ascii-compatible bytes, and I'll report this upstream?

File: hypothesis-python/RELEASE.rst
Comment (jobh) [reply hypothesis-python/RELEASE.rst]: Done. I didn't manage to make the `numpy.typing` xrefs work I'm afraid, not sure if it's a missing intersphinx mapping or what

```
OK - :class:`numpy.ndarray[scalar type] <numpy.ndarray>`
fail :class:`numpy.typing.NDArray[scalar type] <numpy.typing.NDArray>`
```

File: hypothesis-python/src/hypothesis/extra/numpy.py
Comment (jobh) [inline hypothesis-python/src/hypothesis/extra/numpy.py:1112]: This is my attempt at making an arbitrary nested structure of fixed dimension. Is there a (simple) better way?

**Comment by jobh** [comment]
Aside, and FYI, I often (not only on this PR) get failed health checks for unrelated tests, for example
```
  File "/Users/runner/work/hypothesis/hypothesis/hypothesis-python/tests/cover/test_type_lookup.py", line 218, in test_can_generate_from_all_registered_types
```

**Comment by jobh** [comment]
Thanks for your review and comments @Zac-HD , I'm ready for the next round :-)

**Review by Zac-HD — APPROVED** [review]
I'll try pushing a revised release note, and then we're ready to merge!

Thanks again @jobh - you've been shipping some fantastic stuff to our ~half-million users lately 💖

**Comment by jobh** [comment]
Not ndling DTypeLike, sorry

**Comment by jobh** [comment]
> Thanks again @jobh - you've been shipping some fantastic stuff to our ~half-million users lately sparkling_heart

It's certainly a refreshing change from my usual single-digit userbase :-D

Anyway, about my inarticulate comment from earlier: I didn't really think of supporting DTypeLike etc, but it should be fairly easy to add in a followup PR (which would allow that bug to be properly closed).

Also I didn't catch the "shape" part of the release notes, which I'm afraid is correct only in the case of `shape == typing.Any` — I don't think the numpy shape typing proposal is finalized yet? so there's just the placeholder for supporting it.

I guess I'm on the hook now (and happy to be), but not promising anything in the near term as my time budget for this work was blown some time ago and I need to start actually using hypothesis so I can defend that ;-)

**Comment by Zac-HD** [comment]
> I guess I'm on the hook now (and happy to be), but not promising anything in the near term as my time budget for this work was blown some time ago and I need to start actually using hypothesis so I can defend that ;-)

While I and our users will appreciate whatever you want to contribute, I do want to emphasize that you're _not_ 'on the hook' for anything; accepting a commit bit doesn't come with any obligations, and it's pretty important to me personally that Hypothesis maintainers have a culture of only working on things you actually want to!  If I don't feel like reviewing or writing code for free, I just don't.

So go ahead and use Hypothesis without contributing with my blessing, for all the bugs that's worth 🙏

**Pull request merged**
